app/services/notify_service.rb

```diff
@@ -11,126 +11,175 @@ class NotifyService < BaseService
-      WHERE st.mention_id IS NOT NULL AND s.visibility = 3
-    SQL
-  end
+    def initialize(notification)
+      @notification = notification
+      @recipient = notification.account
+      @sender = notification.from_account
+      @policy = NotificationPolicy.find_or_initialize_by(account: @recipient)
+    end
 
-  def from_staff?
-    @notification.from_account.local? && @notification.from_account.user.present? && @notification.from_account.user_role&.overrides?(@recipient.user_role)
-  end
+    def filter?
+      return false if override_for_sender?
 
-  def optional_non_following_and_direct?
-    direct_message? &&
-      @recipient.user.settings['interactions.must_be_following_dm'] &&
-      !following_sender? &&
-      !response_to_recipient?
-  end
+      from_limited? ||
+        (@policy.filter_not_following? && not_following?) ||
+        (@policy.filter_not_followers? && not_follower?) ||
+        (@policy.filter_new_accounts? && new_account?) ||
+        (@policy.filter_private_mentions? && not_following? && private_mention_not_in_response?)
+    end
 
-  def hellbanned?
-    @notification.from_account.silenced? && !following_sender?
-  end
+    private
 
-  def from_self?
-    @recipient.id == @notification.from_account.id
-  end
+    def not_following?
+      !@recipient.following?(@sender)
+    end
 
-  def domain_blocking?
-    @recipient.domain_blocking?(@notification.from_account.domain) && !following_sender?
-  end
+    def not_follower?
+      follow = Follow.find_by(account: @sender, target_account: @recipient)
+      follow.nil? || follow.created_at > NEW_FOLLOWER_THRESHOLD.ago
+    end
 
-  def blocked?
-    blocked   = @recipient.unavailable?
-    blocked ||= from_self? && @notification.type != :poll
+    def new_account?
+      @sender.created_at > NEW_ACCOUNT_THRESHOLD.ago
+    end
 
-    return blocked if message? && from_staff?
+    def override_for_sender?
+      NotificationPermission.exists?(account: @recipient, from_account: @sender)
+    end
+
+    def from_limited?
+      @sender.silenced? && not_following?
+    end
 
-    blocked ||= domain_blocking?
-    blocked ||= @recipient.blocking?(@notification.from_account)
-    blocked ||= @recipient.muting_notifications?(@notification.from_account)
-    blocked ||= hellbanned?
-    blocked ||= optional_non_follower?
-    blocked ||= optional_non_following?
-    blocked ||= optional_non_following_and_direct?
-    blocked ||= conversation_muted?
-    blocked ||= blocked_mention? if @notification.type == :mention
-    blocked
+    def private_mention_not_in_response?
+      @notification.type == :mention && @notification.target_status.direct_visibility? && !response_to_recipient?
+    end
+
+    def response_to_recipient?
+      return false if @notification.target_status.in_reply_to_id.nil?
+
+      Status.count_by_sql([<<-SQL.squish, id: @notification.target_status.in_reply_to_id, recipient_id: @recipient.id, sender_id: @sender.id, depth_limit: 100]).positive?
```

Comment:
This line could be something like `stasuses_with_recipient_responses.positive?` -- or something similar, to give a high level summary of what this large sql block is doing, and then move the whole sql count to it's own method (or move all that to a scope/concern on Status directly)